File: python/arrays-and-hashing/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # Arrays for left direction product and right direction product
        left = [1 for _ in range(len(nums) + 1)]
        right = [1 for _ in range(len(nums) + 1)]

        # 2 separate O(n) loops to populate the arrays in each direction
        for i in range(len(nums)):
            left[i + 1] = left[i] * nums[i]
        for i in range(len(nums) - 1, -1, -1):
            right[i] = right[i + 1] * nums[i]

        # Final return array
        final = [1 for _ in range(len(nums))]

        # Populate the final array based on the multiplication of left and right arrays
        for i in range(len(nums)):
            final[i] = left[i] * right[i + 1]

        return final

File: c++/arrays-and-hashing/product-of-array-except-self.cc
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        std::vector<int> prefix = std::vector<int>{1};
        std::vector<int> suffix = std::vector<int>{1};

        int product = 1;
        for (int i = 0; i < nums.size(); ++i) {
            product = product * nums[i];
            prefix.push_back(product);
        }

        product = 1;
        for (int i = nums.size() - 1; i > -1; --i) {
            product = product * nums[i];
            suffix.push_back(product);
        }
        
        prefix.push_back(1);
        suffix.push_back(1);

        std::reverse(suffix.begin(), suffix.end());

        std::vector<int> answer;
        for (int i = 0; i < nums.size(); ++i) {
            answer.push_back(prefix[i] * suffix[i+2]);
        }
        return answer;
    }
};
